### PR TITLE
Add the Cleanroom

### PR DESCRIPTION
**Pull request by TechLord22**
This PR adds the Cleanroom. It requires many existing recipes to be run inside it, and serves as an HV progression step. 

There are two configs concerning this PR: one to disable the cleanroom, and one to make multiblocks clean when not inside of it. 

The structure has minimum dimensions of 5x5x5, and maximum of 15x15x15. The width and depth must be odd, while the height may be even. The size of the cleanroom is determined in the controller with GUI buttons. The structure requires plascrete in all of the borders, filter casings on the top panel (excluding the very center for the controller), allows up to 4 doors, and has a minimum number of Plascrete or Cleanroom Glass in the structure. This minimum is equal to 75% of the structure's wall's block amount. Any plascrete can be replaced with Cleanroom Glass. 

The cleanroom's interior has a custom predicate which handles machines requiring cleanrooms. The default cleanroom bans any generator, muffler hatch, and any other cleanroom. When a machine meets these conditions, and it is able to "accept" a cleanroom, it is given the cleanroom it is located inside for use in running recipes.

It additionally allows the new Passthrough Item and Fluid hatches in the structure walls. These auto input items or fluids on one side into its inventory, and subsequently auto output out the opposite side. Power can be transmitted inside and out using Diodes or Hulls, which each adopt the proper casing textures in the structure as well. The cleanroom additionally has maintenance, which works as one expects except that the adjustable hatch adjustments make no change. Any hatch is allowed anywhere in the structure excluding the filter casings.

Machines running cleanroom required recipes outside the cleanroom do not run and do not consume inputs at all. Machines running recipes in the cleanroom will work as normal, except that the cleanroom becoming dirty during the recipe will cause progress to reverse. There is no voiding in the cleanroom when not clean. 

The clean states work as follows: it is either clean or not clean, with no in between. Every 5 seconds, the multiblock will increase its progress towards being fully clean. The cleanliness increase is 5% at LV, 10% at MV, 15% at HV, and so on. During this time, if the cleanroom does not recieve enough power for its power draw, it will reverse its progress towards cleaning by subtracting 5% cleanliness. This means that the cleanroom must be continually powered whenever recipes are run inside it. It also constantly consumes power, regardless of whether machines are running inside it or not. The base power drain is 30 EU/t, but drops to 4 when fully cleaned. When there are maintenance problems, the cleanroom's cleanliness increase is reduced by 1% per problem, and its no power decrease is enlarged by 1% for each problem as well.

The cleanroom is designed to be easily extendable by other mods to add their own "cleanrooms." The CleanroomType system was designed in such a way that any amount of cleanrooms which provide their own unique type (or multiple) can be created and used in recipes.

Any RecipeBuilder accepts the cleanroom property, allowing Modpack Devs and Addons to use any machine they wish to inside the cleanroom. CT Support works by specifying a String name of the cleanroom type. An example of this for the default cleanroom: `.property("cleanroom", "cleanroom")` on a RecipeBuilder.

### CODE REVIEW

**Comment by Syrcan** [comment]
I haven't looked at the code yet (I want to save headaches for later), but here's what I can tell from the mechanics in-game:
- The 30EU/t drain is permanent, where in GT5u it would go down to 4EU/t or something once fully clean.
- I have serious concerns about gating, as this is currently as hard (if not harder) than GT5u. The plascrete recipe was agreed to be awful. MV would become extremely bland without both tiers of circuits. I would personally like for the cleanroom to be required for Workstation, but not the previous circuits.
- I like the passthrough hatches. I hope automating the machines in the cleanroom won't be a pain for players.
- Doors can be kept open without any downsides.
- The progression for cleanliness should remain at 100% instead of looping back to 0. This will make lack of energy or opening doors less awkward.
- Other notes would be related to the logic having to be rewritten from scratch / copy pasted, which doesn't implement everything correctly from AbstractRecipeLogic, but I guess that's gonna be for another PR.

**Comment by ChromaPIE** [comment]
So uh, it requires 60% of its blocks to be plascrete, so a 7x7x7 cleanroom is actaully not possible?

File: src/main/java/gregtech/common/metatileentities/multi/electric/MetaTileEntityCleanroom.java [hunk 1]
```diff
@@ -0,0 +1,533 @@
...
+        center[height - 1] = C;
+
+        // central sections
+        for (int i = 1; i < height - 1; i++) {
+            frontBack[i] = W;
+            inner[i] = I;
+            center[i] = I;
+        }
+
+        TraceabilityPredicate casing = states(getCasingState()).setMinGlobalLimited(width * height * depth * 3 / 5)
```
Comment (ChromaPIE) [inline src/main/java/gregtech/common/metatileentities/multi/electric/MetaTileEntityCleanroom.java:166]:
```suggestion
        TraceabilityPredicate casing = states(getCasingState()).setMinGlobalLimited((width * height * depth - (width - 2) * (height - 2) * (depth - 2)) * 3 / 5)
```

Comment (ChromaPIE) [inline src/main/java/gregtech/common/metatileentities/multi/electric/MetaTileEntityCleanroom.java:166]:
```suggestion
        TraceabilityPredicate casing = states(getCasingState()).setMinGlobalLimited((width * height * depth - (width - 2) * (height - 2) * (depth - 2)) * 3 / 5)
```

I believe its min limit was designed to be 60% of the **wall**, instead of the whole space it takes including the hollow part, or things won't be possible.

**Review by ChromaPIE — CHANGES_REQUESTED** [review]

**Comment by DStrand1** [comment]
I added a commit which adds channels in hulls. At the moment, it is partially broken, as I am not calling the read/write NBT methods at the right time for the ME network. @PrototypeTrousers guidance on this issue (and general reviews of the commit d7a8258) would be greatly appreciated

**Comment by ChromaPIE** [comment]
There be a log spam spamming
```
[22:44:35] [Server thread/WARN] [GregTech]: There is no property with key cleanroom
[22:44:35] [Server thread/WARN] [GregTech]: Stacktrace:
java.lang.IllegalArgumentException: null
	at gregtech.api.recipes.recipeproperties.RecipePropertyStorage.getRecipePropertyValue(RecipePropertyStorage.java:90) [RecipePropertyStorage.class:?]
	at gregtech.api.recipes.Recipe.getProperty(Recipe.java:588) [Recipe.class:?]
	at gregtech.api.capability.impl.AbstractRecipeLogic.checkRecipe(AbstractRecipeLogic.java:311) [AbstractRecipeLogic.class:?]
	at gregtech.api.capability.impl.MultiblockRecipeLogic.checkRecipe(MultiblockRecipeLogic.java:270) [MultiblockRecipeLogic.class:?]
	at gregtech.api.capability.impl.AbstractRecipeLogic.trySearchNewRecipe(AbstractRecipeLogic.java:289) [AbstractRecipeLogic.class:?]
	at gregtech.api.capability.impl.MultiblockRecipeLogic.trySearchNewRecipeCombined(MultiblockRecipeLogic.java:130) [MultiblockRecipeLogic.class:?]
	at gregtech.api.capability.impl.MultiblockRecipeLogic.trySearchNewRecipe(MultiblockRecipeLogic.java:122) [MultiblockRecipeLogic.class:?]
	at gregtech.api.capability.impl.AbstractRecipeLogic.update(AbstractRecipeLogic.java:146) [AbstractRecipeLogic.class:?]
	at gregtech.api.capability.impl.MultiblockRecipeLogic.updateWorkable(MultiblockRecipeLogic.java:38) [MultiblockRecipeLogic.class:?]
	at gregtech.api.metatileentity.multiblock.RecipeMapMultiblockController.updateFormedValid(RecipeMapMultiblockController.java:141) [RecipeMapMultiblockController.class:?]
	at gregtech.api.metatileentity.multiblock.MultiblockControllerBase.update(MultiblockControllerBase.java:84) [MultiblockControllerBase.class:?]
	at gregtech.api.metatileentity.MetaTileEntityHolder.update(MetaTileEntityHolder.java:162) [MetaTileEntityHolder.class:?]
	at net.minecraft.world.World.redirect$zcc000$tileEntityUpdate(World.java:4137) [amu.class:?]
	at net.minecraft.world.World.updateEntities(World.java:1838) [amu.class:?]
	at net.minecraft.world.WorldServer.updateEntities(WorldServer.java:613) [oo.class:?]
	at net.minecraft.server.MinecraftServer.updateTimeLightAndEntities(MinecraftServer.java:767) [MinecraftServer.class:?]
	at net.minecraft.server.MinecraftServer.tick(MinecraftServer.java:668) [MinecraftServer.class:?]
	at net.minecraft.server.integrated.IntegratedServer.tick(IntegratedServer.java:279) [chd.class:?]
	at net.minecraft.server.MinecraftServer.run(MinecraftServer.java:526) [MinecraftServer.class:?]
	at java.lang.Thread.run(Unknown Source) [?:1.8.0_281]
```

File: src/main/java/gregtech/loaders/recipe/CircuitRecipes.java [hunk 2]
```diff
@@ -67,125 +68,133 @@ private static void waferRecipes() {
...
+        CUTTER_RECIPES.recipeBuilder().duration(900).EUt(VA[IV]).input(HIGH_POWER_INTEGRATED_CIRCUIT_WAFER).output(HIGH_POWER_INTEGRATED_CIRCUIT, 2).cleanroom(CleanroomType.CLEANROOM).buildAndRegister();
+        CUTTER_RECIPES.recipeBuilder().duration(900).EUt(VA[LuV]).input(ULTRA_HIGH_POWER_INTEGRATED_CIRCUIT_WAFER).output(ULTRA_HIGH_POWER_INTEGRATED_CIRCUIT, 2).cleanroom(CleanroomType.CLEANROOM).buildAndRegister();
+        CUTTER_RECIPES.recipeBuilder().duration(900).EUt(192).input(NOR_MEMORY_CHIP_WAFER).output(NOR_MEMORY_CHIP, 16).cleanroom(CleanroomType.CLEANROOM).buildAndRegister();
+        CUTTER_RECIPES.recipeBuilder().duration(900).EUt(192).input(NAND_MEMORY_CHIP_WAFER).output(NAND_MEMORY_CHIP, 32).cleanroom(CleanroomType.CLEANROOM).buildAndRegister();
+        CUTTER_RECIPES.recipeBuilder().duration(900).EUt(VA[MV]).input(CENTRAL_PROCESSING_UNIT_WAFER).output(CENTRAL_PROCESSING_UNIT, 8).cleanroom(CleanroomType.CLEANROOM).buildAndRegister();
+        CUTTER_RECIPES.recipeBuilder().duration(900).EUt(64).input(INTEGRATED_LOGIC_CIRCUIT_WAFER).output(INTEGRATED_LOGIC_CIRCUIT, 8).cleanroom(CleanroomType.CLEANROOM).buildAndRegister();
```
Comment (ChromaPIE) [inline src/main/java/gregtech/loaders/recipe/CircuitRecipes.java:196]:
```suggestion
        CUTTER_RECIPES.recipeBuilder().duration(900).EUt(64).input(INTEGRATED_LOGIC_CIRCUIT_WAFER).output(INTEGRATED_LOGIC_CIRCUIT, 8).CLEANROOM).buildAndRegister();
```
This recipe requiring a cleanroom causes fatal deadlock by making it impossible to make the very first HV circuit, which is required for a cleanroom controller.

**Review by ChromaPIE — CHANGES_REQUESTED** [review]

File: src/main/java/gregtech/common/metatileentities/electric/MetaTileEntityHull.java [hunk 3]
```diff
@@ -81,4 +87,30 @@ public void addInformation(ItemStack stack, @Nullable World player, List<String>
...
+        if (isFirstTick()) {
+            getProxy();
+        }
+    }
+
+    @Nonnull
+    @Override
+    @Optional.Method(modid = GTValues.MODID_APPENG)
+    public AECableType getCableConnectionType(@Nonnull AEPartLocation part) {
+        return AECableType.SMART;
```
Comment (ChromaPIE) [inline src/main/java/gregtech/common/metatileentities/electric/MetaTileEntityHull.java:103]: Should find a way to allow players have the channel gameplay displayed to take advantage of this feature, or they might have to get AE2 stuff.

**Comment by ChromaPIE** [comment]
> I added a commit which adds channels in hulls. At the moment, it is partially broken, as I am not calling the read/write NBT methods at the right time for the ME network. @PrototypeTrousers guidance on this issue (and general reviews of the commit [d7a8258](https://github.com/GregTechCEu/GregTech/commit/d7a825874040163ea6627eb2638e97ea84ef4b3d)) would be greatly appreciated

JEI integration would be completely broken if AE2 is not present.

File: src/main/java/gregtech/common/metatileentities/multi/electric/MetaTileEntityCleanroom.java (hunk 1 above)
Comment (TechLord22) [reply src/main/java/gregtech/common/metatileentities/multi/electric/MetaTileEntityCleanroom.java:166]: It is supposed to be the entire volume of the cleanroom, not just the walls. Doing this keeps the amount of required plascrete at an amount that does not allow mass usage of hatches instead of the intended casings. 

Additionally, this is intended to be applied to exclusively the plascrete and not the glass. It cannot be feasibly changed either due to the nature of the multiblock system.

**Comment by ChromaPIE** [comment]
It worths mentioning that in Gregicality Multiblocks the large engraver has a muffler hatch in its structure, it means that engravers would never be able to process cleanroom recipes cuz it can't meet both requirements.

**Comment by Quarri6343** [comment]
I found several issues while having fun with it:
・it does not ban Primitive Blast Furnace which releases so many smoke particles
・Cleanroom controller has voiding mode
・you can stuck in the wall when auto build the multiblock (seems like the Multiblock Helper App's bug)
・multiblock can be formed with wooden doors (sorry if it is a feature. however it is still very weird for a GT5u player)
・if you are in the very late game, you have to put a 12 length Assembly Line into the 13 width Cleanroom which is not beautiful in some cases

**Comment by ChromaPIE** [comment]
Would be even better to introduce a specific door only for the Cleanroom using a texture with no, you know, **holes** in it.

**Comment by TechLord22** [comment]
With my latest commit, [85ac6eb](https://github.com/GregTechCEu/GregTech/pull/618/commits/85ac6ebcc6e981615077f597335b2e35431dda3d), the following has been changed:

Primitive Multiblocks and Drills are now blacklisted from forming the cleanroom structure.
The voiding mode button is now no longer present in the controller.
All assemblyline recipes requiring the cleanroom have been removed. This was not intended.

**Comment by Quarri6343** [comment]
I messed with it again and found the following bugs:

1. Blacklight recipe has wrong tier glass
it is IV~ZPM tier so there should be Laminated Glass or Fusion Glass
2. Cleanroom structure preview has 2 missing blocks
キャプチャ1250
3. Cleanroom structure preview always shows the 5x5x5 Cleanroom regardless of the size settings in the Cleanroom GUI
Central Monitor changes its preview height when the height setting was edited so it is a bug
4. Cleanroom can be rotated with a wrench 
I found this because the missing blocks in the structure preview rotated
5. When you auto build the Cleanroom in the creative mode, tons of fluid passthrough hatch will appear
キャプチャ1251
6. Debug function of the Multiblock Helper app always throws the same error regardless of whether the Cleanroom is formed or not
キャプチャ1252
7. Cleanroom forms with misaligned doors
キャプチャ1253
8. You can place as many vanilla torches which emit smoke particle as you want
9. Primitive Machinery still can be placed in the functioning cleanroom
10. You can attach Automatic Filtration Maintenance Hatch to the Cleanroom (it is recursive!)
11. Sterilizing Filter Casing requires Tritanium which is the UV material
this can block the normal progression since the Wetware Circuit is intended to be made in ZPM age
12. Multiblock which is no longer in the clean state displays "Machine needs more energy" although it has enough power
it should display "Machine is contaminated!" or something which indicates the proper machine state
キャプチャ1254
13. Automatic Filtration Maintenance Hatch does not work in the Cleanroom (also outside?)
when the Cleanroom stopped working, the multiblock stops even if it has the AFMH

File: src/main/java/gregtech/api/capability/impl/CleanroomLogic.java [hunk 4]
```diff
@@ -0,0 +1,228 @@
...
+
+    protected boolean consumeEnergy(boolean simulate) {
+        return ((ICleanroomProvider) metaTileEntity).drainEnergy(simulate);
+    }
+
+    /**
+     * @return true if the cleanroom is able to drain energy, else false
+     */
+    protected boolean hasEnoughPower() {
+        if (!consumeEnergy(true)) {
```
Comment (ALongStringOfNumbers) [inline src/main/java/gregtech/api/capability/impl/CleanroomLogic.java:79]: If my previous suggestion is taken, this check would not be needed. Also, it does not quite make sense. Right now the structure always drains energy, even if it may not have enough energy in the first place, but then does a simulate check here?

File: src/main/java/gregtech/api/capability/impl/CleanroomLogic.java [hunk 5]
```diff
@@ -0,0 +1,228 @@
...
+            if (progressTime >= 2) {
+                if (ConfigHolder.machines.recipeProgressLowEnergy) this.progressTime = 1;
+                else this.progressTime = Math.max(1, progressTime - 2);
+
+                hasNotEnoughEnergy = true;
+            }
+            return false;
+        }
+
+        if (this.hasNotEnoughEnergy && ((ICleanroomProvider) metaTileEntity).getEnergyInputPerSecond() > 19L * GTValues.VA[((ICleanroomProvider) metaTileEntity).getEnergyTier()]) {
```
Comment (ALongStringOfNumbers) [inline src/main/java/gregtech/api/capability/impl/CleanroomLogic.java:89]: Why 19?

File: src/main/java/gregtech/api/capability/impl/CleanroomLogic.java [hunk 6]
```diff
@@ -0,0 +1,228 @@
...
+            }
+        }
+    }
+
+    /**
+     * @param workingEnabled the new state of the cleanroom's ability to work: true to change to enabled, else false
+     */
+    public void setWorkingEnabled(boolean workingEnabled) {
+        this.isWorkingEnabled = workingEnabled;
+        metaTileEntity.markDirty();
```
Comment (ALongStringOfNumbers) [inline src/main/java/gregtech/api/capability/impl/CleanroomLogic.java:121]: This also needs to be synced via writeCustomData, otherwise we will have the issue that we had with the miners not properly syncing stuff

File: src/main/java/gregtech/api/recipes/builders/BlastRecipeBuilder.java [hunk 7]
```diff
@@ -32,12 +34,12 @@ public BlastRecipeBuilder copy() {
...
 
     @Override
-    public boolean applyProperty(String key, Object value) {
+    public boolean applyProperty(@Nonnull String key, Object value) {
         if (key.equals(TemperatureProperty.KEY)) {
             this.blastFurnaceTemp(((Number) value).intValue());
             return true;
         }
-        return true;
+        return super.applyProperty(key, value);
```
Comment (ALongStringOfNumbers) [inline src/main/java/gregtech/api/recipes/builders/BlastRecipeBuilder.java:42]: This is a behavior change. Previously this method would return true, now if there is no cleanroom property in the recipe, this will return false.

Although, that may be fine? I am not sure if this is supposed to return `true` regardless of if the property was applied, or false like all the other specialized builders.

File: src/main/java/gregtech/api/recipes/RecipeBuilder.java [hunk 8]
```diff
@@ -95,9 +102,27 @@ protected RecipeBuilder(RecipeBuilder<R> recipeBuilder) {
         this.EUt = recipeBuilder.EUt;
         this.hidden = recipeBuilder.hidden;
         this.onBuildAction = recipeBuilder.onBuildAction;
+        this.cleanroom = recipeBuilder.cleanroom;
+    }
+
+    public R cleanroom(@Nullable CleanroomType cleanroom) {
```
Comment (ALongStringOfNumbers) [inline src/main/java/gregtech/api/recipes/RecipeBuilder.java:108]: If this is how you are doing this, `CleanroomType` needs to be exposed to Crafttweaker, or you need to have a method in CTRecipeBuilder that takes string `CleanroomType` keys and then passes the `CleanroomType` object to the backingBuilder.

File: src/main/java/gregtech/common/ConfigHolder.java [hunk 9]
```diff
@@ -95,6 +95,14 @@ public static class MachineOptions {
                 "Add the unlocalized Recipe Map name to blacklist the machine.",
                 "Default: All machines allowed"})
         public String[] processingArrayBlacklist = new String[0];
+
+        @Config.Comment({"Whether to enable the cleanroom, required for various recipes.", "Default: true"})
+        public boolean enableCleanroom = true;
+
+        @Config.Comment({"Whether multiblocks are able to run cleanroom recipes outside the cleanroom.",
+                "This does nothing is B:enableCleanroom is false.",
```
Comment (ALongStringOfNumbers) [inline src/main/java/gregtech/common/ConfigHolder.java:103]: "does nothing `if` ..."

File: src/main/java/gregtech/api/capability/impl/CleanroomLogic.java [hunk 10]
```diff
@@ -0,0 +1,228 @@
...
+        return this.progressTime;
+    }
+
+    public int getMaxProgress() {
+        return MAX_PROGRESS;
+    }
+
+    public double getProgressPercent() {
+        return getProgressTime() * 1.0 / getMaxProgress();
+    }
```
Comment (ALongStringOfNumbers) [inline src/main/java/gregtech/api/capability/impl/CleanroomLogic.java:151]: Why is this a double when both parameters are int, and the only place that uses this method casts it back to an int as well?

File: src/main/java/gregtech/common/metatileentities/electric/MetaTileEntityDiode.java [hunk 11]
```diff
@@ -23,16 +26,20 @@
...
 
+import javax.annotation.Nonnull;
 import javax.annotation.Nullable;
 import java.util.Arrays;
 import java.util.List;
 
 import static gregtech.api.capability.GregtechDataCodes.AMP_INDEX;
 
-public class MetaTileEntityDiode extends TieredMetaTileEntity {
+public class MetaTileEntityDiode extends MetaTileEntityMultiblockPart implements EnergyContainerHandler.IEnergyChangeListener {
```
Comment (ALongStringOfNumbers) [inline src/main/java/gregtech/common/metatileentities/electric/MetaTileEntityDiode.java:40]: Why was this changed to implement this Energy Change listener when it does not do anything with it?

In fact, this interface seems unused entirely, and from commit history was last used to update comparator values for machines, which was removed in 2019 in CE.

**Review by ALongStringOfNumbers — COMMENTED** [review]
Just an idea, how hard would it be to increase the time required for the cleanroom to become fully clean based on the internal size of the cleanroom?

Also, instead of decreasing the cleanliness by 1% per maintance issue, what do you think about decreasing cleanliness by 1% per maintenance issue, per cleanroom power draw cycle?

**Comment by Quarri6343** [comment]
I  found the following bugs in my second review:
1. progress persentage in the controller ui is bugged
キャプチャ1346
2. Cleanroom Glass has 3,000 blest resistance value (it is the issue across all reinforced glasses)
3. Machine Hull drains energy from the ME network (it shouldn't because it only works as a cable)
4. Cleanroom Controller do animations even in the paused state
5. Putting Central Monitor inside the cleanroom will crash the game (see discord for details)

File: src/main/java/gregtech/api/capability/impl/CleanroomLogic.java [hunk 12]
```diff
@@ -0,0 +1,245 @@
...
+        if (!this.isActive) setActive(true);
+
+        // increase progress
+        progressTime++;
+        if (progressTime % getMaxProgress() != 0) return;
+        progressTime = 0;
+
+        int amountToClean;
+        if (hasEnoughPower()) amountToClean = BASE_CLEAN_AMOUNT * (getTierDifference() + 1);
+        else amountToClean = -BASE_CLEAN_AMOUNT;
```
Comment (Quarri6343) [inline src/main/java/gregtech/api/capability/impl/CleanroomLogic.java:72]: This line is impossible to reach regardless  of hasEnoughPower() because the progress cannot be lower than 1(see line 55 and 56)
This makes cleanroom be able to stay clean without power supply
キャプチャ1379

File: src/main/java/gregtech/api/capability/impl/CleanroomLogic.java [hunk 13]
```diff
@@ -0,0 +1,245 @@
...
+
+    /**
+     * @return true if the cleanroom is able to drain energy, else false
+     */
+    protected boolean hasEnoughPower() {
+        if (this.hasNotEnoughEnergy && ((ICleanroomProvider) metaTileEntity).getEnergyInputPerSecond() > 19L * GTValues.VA[((ICleanroomProvider) metaTileEntity).getEnergyTier()]) {
+            this.hasNotEnoughEnergy = false;
+        }
+
+        return true;
```
Comment (Quarri6343) [inline src/main/java/gregtech/api/capability/impl/CleanroomLogic.java:91]: `true if the cleanroom is able to drain energy, else false`
it always returns true

File: src/main/java/gregtech/api/metatileentity/multiblock/ICleanroomProvider.java [hunk 14]
```diff
@@ -0,0 +1,51 @@
...
+     */
+    boolean drainEnergy(boolean simulate);
+
+    /**
+     * @return the amount of energy input per second
+     */
+    long getEnergyInputPerSecond();
+
+    /**
+     * @return the tier {@link gregtech.api.GTValues.V} of energy the cleanroom uses at minimum
```
Comment (Quarri6343) [inline src/main/java/gregtech/api/metatileentity/multiblock/ICleanroomProvider.java:48]: Idea throws error at `.V`
I recommend to write like `@link gregtech.api.GTValues#V` ...

File: src/main/java/gregtech/common/ConfigHolder.java [hunk 15]
```diff
@@ -95,6 +95,14 @@ public static class MachineOptions {
                 "Add the unlocalized Recipe Map name to blacklist the machine.",
                 "Default: All machines allowed"})
         public String[] processingArrayBlacklist = new String[0];
+
+        @Config.Comment({"Whether to enable the cleanroom, required for various recipes.", "Default: true"})
+        public boolean enableCleanroom = true;
+
+        @Config.Comment({"Whether multiblocks are able to run cleanroom recipes outside the cleanroom.",
```
Comment (Quarri6343) [inline src/main/java/gregtech/common/ConfigHolder.java:102]: I think it is better to take Automatic Filteration Maintenance Hatch into account.
cf: `Whether multiblocks are able to run cleanroom recipes outside the cleanroom without any requirement.`

**Review by Quarri6343 — COMMENTED** [review]
we need more reviewers...

File: src/main/java/gregtech/loaders/recipe/chemistry/DistillationRecipes.java [hunk 16]
```diff
@@ -94,6 +95,7 @@ public static void init() {
         DISTILLATION_RECIPES.recipeBuilder()
                 .fluidInputs(SeedOil.getFluid(24))
                 .fluidOutputs(Lubricant.getFluid(12))
+                .cleanroom(CleanroomType.CLEANROOM)
```
Comment (Quarri6343) [inline src/main/java/gregtech/loaders/recipe/chemistry/DistillationRecipes.java:98]: why Seed Oil distillation requires cleanroom...?
If it is intended, why Fish Oil distillation does not require cleanroom?

**Review by Quarri6343 — COMMENTED** [review]
Someone said cleanroom can be unformed by reloading the world in discord dev channel and I confirmed it happens in the latest commit
However, I don't know which line causes it

**Review by Quarri6343 — APPROVED** [review]

File: src/main/java/gregtech/common/metatileentities/multi/electric/MetaTileEntityCleanroom.java [hunk 17]
```diff
@@ -0,0 +1,688 @@
...
+
+        String[] slice = new String[hDist + 1]; // "BBBBB", "X   X", "X   X", "X   X", "BFFFB"
+        Arrays.fill(slice, insideBuilder.toString());
+        slice[0] = borderBuilder.toString();
+        slice[slice.length - 1] = roofBuilder.toString();
+
+        String[] center = Arrays.copyOf(slice, slice.length); // "BBBBB", "X   X", "X   X", "X   X", "BFSFB"
+        center[center.length - 1] = controllerBuilder.toString();
+
+        TraceabilityPredicate wallPredicate = states(getCasingState(), getGlassState()).setMinGlobalLimited((((lDist + rDist + 1) * (bDist + fDist + 1) * (hDist + 1)) - ((lDist + rDist - 1) * (bDist + fDist - 1) * (hDist - 1))) * 3 / 4);
```
Comment (ALongStringOfNumbers) [inline src/main/java/gregtech/common/metatileentities/multi/electric/MetaTileEntityCleanroom.java:294]: The use of casing and glass states means that `borderBuilder` can be made of casings or glass, which means that you can have glass along the edges of the structure, which I am not sure is intended or not.

**Review by ALongStringOfNumbers — COMMENTED** [review]
Found a funny issue with the JEI preview for the cleanroom. Since the door is in 2 block spaces, is is shown that 2 doors are required in the JEI components list.

File: src/main/java/gregtech/common/metatileentities/multi/electric/MetaTileEntityCleanroom.java [hunk 18]
```diff
@@ -0,0 +1,688 @@
...
+    }
+
+    /**
+     * @param world     the world to check
+     * @param pos       the pos to check and move
+     * @param direction the direction to move
+     * @return if a block is a valid wall block at pos moved in direction
+     */
+    public boolean isBlockWall(@Nonnull World world, @Nonnull BlockPos.MutableBlockPos pos, @Nonnull EnumFacing direction) {
+        return MetaBlocks.CLEANROOM_CASING.getState(BlockCleanroomCasing.CasingType.PLASCRETE).equals(world.getBlockState(pos.move(direction)));
```
Comment (ALongStringOfNumbers) [inline src/main/java/gregtech/common/metatileentities/multi/electric/MetaTileEntityCleanroom.java:210]: Shouldn't this also check if the block is a Cleanroom Glass?

**Review by ALongStringOfNumbers — APPROVED** [review]
